programs/mango-v4/src/instructions/account_settle_fees_accrued_with_mngo.rs

```diff
@@ -0,0 +1,98 @@
+) -> Result<()> {
+    // Cannot settle with yourself
+    require_keys_neq!(
+        ctx.accounts.account.key(),
+        ctx.accounts.dao_account.key(),
+        MangoError::CannotSettleWithSelf
+    );
+
+    let group = ctx.accounts.group.load()?;
+
+    let mut mngo_bank = ctx.accounts.mngo_bank.load_mut()?;
+    let mut settle_bank = ctx.accounts.settle_bank.load_mut()?;
+
+    let mut account = ctx.accounts.account.load_full_mut()?;
+    // account constraint #1
+    require!(
+        account.fixed.is_owner_or_delegate(ctx.accounts.owner.key()),
+        MangoError::SomeError
+    );
+    let mut dao_account = ctx.accounts.dao_account.load_full_mut()?;
+
+    // check positions exist, for nicer error messages
+    {
+        account.token_position(mngo_bank.token_index)?;
+        account.token_position(settle_bank.token_index)?;
+        dao_account.token_position(mngo_bank.token_index)?;
+        dao_account.token_position(settle_bank.token_index)?;
+    }
+
+    let discount_rate = I80F48::from_num::<f32>(group.fees_mngo_discount_rate);
+    let now_ts = Clock::get()?.unix_timestamp.try_into().unwrap();
+
+    let mut max_settle = I80F48::from_num::<u64>(max_settle.min(account.fixed.fees_accrued));
+
+    // if mngo token position has borrows, skip settling
+    let account_mngo_token_position = account.token_position_mut(mngo_bank.token_index)?.0;
+    let account_mngo_native = account_mngo_token_position.native(&mngo_bank);
+    if account_mngo_native.is_negative() {
+        msg!(
+            "mngo token position ({} native mngo) has borrows, nothing will be settled",
+            account_mngo_native
+        );
+        return Ok(());
+    }
+    let mngo_oracle_price = mngo_bank.oracle_price(
+        &AccountInfoRef::borrow(&ctx.accounts.mngo_oracle.as_ref())?,
+        None,
+    )?;
+    // mngo is exchanged at a discount
+    let mut max_settle_mngo = cm!(max_settle * discount_rate / mngo_oracle_price);
+    // settlement is restricted to accounts token position
+    max_settle_mngo = max_settle_mngo.min(account_mngo_native);
+    max_settle = cm!(max_settle_mngo * mngo_oracle_price / discount_rate);
+
+    // move mngo from user to dao
+    let dao_mngo_token_position = dao_account.token_position_mut(mngo_bank.token_index)?.0;
+    mngo_bank.withdraw_without_fee(
+        account_mngo_token_position,
+        max_settle_mngo,
+        now_ts,
+        mngo_oracle_price,
+    )?;
+    mngo_bank.deposit(dao_mngo_token_position, max_settle_mngo, now_ts)?;
+
+    // move settlement tokens from dao to user
+    let account_settle_token_position = account.token_position_mut(settle_bank.token_index)?.0;
+    let dao_settle_token_position = dao_account.token_position_mut(settle_bank.token_index)?.0;
+    settle_bank.withdraw_without_fee(
+        dao_settle_token_position,
+        max_settle,
+        now_ts,
+        mngo_oracle_price,
+    )?;
+    settle_bank.deposit(account_settle_token_position, max_settle, now_ts)?;
```

Comment:
these four balance changes might implicitly close any of these four token positions, you need to track the return values and possibly close deactivate_token_position_... :(

Reply:
would keeping these positions open be an issue?

Reply:
Not directly. The main two reasons are consistency - the other instructions close token positions that were dusted, not doing so is kind of an invalid state. And secondly if the user's MNGO token position stays open, then the user needs a workaround to close it when he wants to reuse the token index (something like depositing 1 native and withdrawing it again)

But yeah, I'm pretty annoyed about having to do this in all instructions that deal with token positions. Possibly it'd be better to skip it and auto-close zeroed token positions in enforce_token_position? That also has some implications for which banks/oracles need to be send for health checks though.

Reply:
I have now deactivated the token positions if not in_use after withdraw, please check latest commit, afaics positions where deposit is happening don't need to be checked

I have also added an additional check which dao fees token position >= max_buyback

Reply:
I have also added deactivations for deposits, to just be consistent